// @sequelize/mysql/lib/_internal/connection-options.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var connection_options_exports = {};
__export(connection_options_exports, {
  BOOLEAN_CONNECTION_OPTION_NAMES: () => BOOLEAN_CONNECTION_OPTION_NAMES,
  CONNECTION_OPTION_NAMES: () => CONNECTION_OPTION_NAMES,
  NUMBER_CONNECTION_OPTION_NAMES: () => NUMBER_CONNECTION_OPTION_NAMES,
  STRING_CONNECTION_OPTION_NAMES: () => STRING_CONNECTION_OPTION_NAMES
});
module.exports = __toCommonJS(connection_options_exports);
var import_utils = require("@sequelize/utils");
const STRING_CONNECTION_OPTION_MAP = {
  charset: void 0,
  database: void 0,
  host: void 0,
  localAddress: void 0,
  password: void 0,
  password1: void 0,
  password2: void 0,
  password3: void 0,
  passwordSha1: void 0,
  socketPath: void 0,
  ssl: void 0,
  user: void 0
};
const STRING_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  STRING_CONNECTION_OPTION_MAP
);
const BOOLEAN_CONNECTION_OPTION_MAP = {
  compress: void 0,
  trace: void 0,
  enableKeepAlive: void 0,
  isServer: void 0,
  insecureAuth: void 0,
  multipleStatements: void 0,
  waitForConnections: void 0
};
const BOOLEAN_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  BOOLEAN_CONNECTION_OPTION_MAP
);
const NUMBER_CONNECTION_OPTION_MAP = {
  port: void 0,
  connectionLimit: void 0,
  connectTimeout: void 0,
  charsetNumber: void 0,
  maxIdle: void 0,
  queueLimit: void 0,
  idleTimeout: void 0,
  maxPreparedStatements: void 0,
  keepAliveInitialDelay: void 0
};
const NUMBER_CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)(
  NUMBER_CONNECTION_OPTION_MAP
);
const CONNECTION_OPTION_NAMES = (0, import_utils.getSynchronizedTypeKeys)({
  ...STRING_CONNECTION_OPTION_MAP,
  ...BOOLEAN_CONNECTION_OPTION_MAP,
  ...NUMBER_CONNECTION_OPTION_MAP,
  infileStreamFactory: void 0,
  flags: void 0,
  authSwitchHandler: void 0,
  connectAttributes: void 0,
  authPlugins: void 0,
  debug: void 0,
  stream: void 0
});
